Police api and navigator

OrianP opened the pull request:
Hi Ray, 

I created two functions to do the following:
1. Fetch the crime data from the police API
2. Get user coordinates using the geolocation (which apparently is another API!) `getCurrentPosition()` method, and then de-structure those coordinates into 'lat' and 'lng' variables which are then interpolated into a url template like so:
`const url = `https://data.police.uk/api/crimes-at-location?&lat=${lat}&lng=${lng}

Each function works individually, however getting the coordinates is an async operation and it takes a few seconds to retrieve those coordinates. Therefore I have to write the code so that the `getCrimes` function is executed only once the coordinates are retrieved. I tried constructing a new promise on line 8, to then call the `getCrimes` function when the url with the lat and lng is ready. This does not seem to be working. You will see the url logged in the console from within the `getUserCoords` function on line 18. But this url is not being passed into the `positionPromise` on line 11. 

Do you have any ideas? 💡 
Feeling quite stuck on this right now. Please review the code and let me know what you think! If you can leave comments on the code on specific lines that would be very helpful!

OrianP committed:
Setup police-api branch

OrianP committed:
Get crimes
create function to fetch crime data at street location

OrianP committed:
Navigator
create function to get user coords from navigator object

raythompsonwebdev reviewed (commented):
HI Orian.  
Here is my review of the code.
 1. getUserCoords function is outside of getCurrentPosition function so function not picking up position argument I think, because of scope. getUserCoords function expression passed into getCurrentposition has the value of the url returned from the function not the position object with coordinates. 
2. I would suggest passing in an anonymous function as callback getCurentPosition so you can pass position object as argument to getCurentPosition.
3. You will need add resolve and reject arguments to promise.then check if response is ok with response.ok and if/else. 
4. If true pass result from fetch (whether you asign it a variable or return it from a function) as argument to resolve() function.  you can apply .json() at this stage to result to get json data and return resolve.
5. if false pass an [error object]([https://developer.mozilla.org/en-US/docs/web/javascript/reference/global_objects/error]) to reject() function to display error and return reject().
6. Once data has been returned from promise as json data add another .then to access the data.

OrianP committed:
Fix bug and loader example
Place navigator.geolocation in IIFE and call getCrimes within it. Add example of how loader might work with Ray's code

OrianP commented:
Hi Ray! 

I managed to get my code to work by calling the `getCrimes` function within an IIFE after the geolocation coordinates are returned. I looked at how you structured your code and followed those principles. Thank you! 🙏 

I understand though that my code might be redundant at this point, as you have already fetched the police API data based on user coordinates in your code, but maybe we can still integrate some of my named functions into your structure when we merge? 

Looking into displaying a loading element - I added a very (ugly) simple example of one way this might work in JS. I copy- pasted your code for displaying the crime panel and added the loader into that. Have a look from line 16 onwards in the script file. I can open a new branch once we've merged the current ones and work on making the loader prettier and animated.

raythompsonwebdev reviewed (approved):
Hi Orian. Code looks great. Glad you got it working.
